update sitemap to support replicated version

reggie7 opened the pull request:
[603](https://github.com/headwirecom/peregrine-cms/issues/603)

reggie7 committed:
Extract PerConstants.PUBLISHED_LABEL

reggie7 committed:
Extract new Page constructor

reggie7 committed:
Add jcr:content PerUtil methods

reggie7 committed:
Fix camel case typo

reggie7 committed:
Use Published version usage in site map

reggie7 committed:
Fix broken test

GastonGonzalez commented on platform/base/core/src/main/java/com/peregrine/sitemap/impl/DefaultSiteMapExtractor.java:
```diff
@@ -147,4 +156,25 @@ protected SiteMapUrlBuilder getUrlBuilder() {
+            final Node node = version.getFrozenNode();
+            return new Page(page, resourceResolver.getResource(node.getPath()));
+        } catch (final RepositoryException e) {
+            return null;
```
@reggie7 - Please add appropriate log level and message.

GastonGonzalez replied:
Looks good. In the future, you can use substitutions in the logger. I believe the method signatures supports up two substitutions.

```
logger.trace("Unable to grab the published version of path: {} ", path, e);
```

reggie7 replied:
Done.

GastonGonzalez commented:
@reggie7 - Please address #608 in this PR. It really should be batched with this change.

reggie7 committed:
Trace the exception

reggie7 committed:
Use substitution in logger.trace

reggie7 committed:
Extend createVersion to accept multiple labels

reggie7 committed:
Add version labels on Site Setup Replication

reggie7 committed:
Fix labels display

reggie7 commented:
> @reggie7 - Please address #608 in this PR. It really should be batched with this change.

Done.

reggie7 commented:
@cmrockwell, I'd like to ask your opinion here. This PR will most likely cause a discrepancy between sitemap rendering in an author instance and publish (with `remote` replication scenario). Namely, since we are actively grabbing the **Published** version from history - it's all inside the author instance only, right? We're not propagating the version history to publish instance.
We could instead e.g. send the final `sitemap.xml` file directly to publish and cache it there as-is on author. What do you think?

reusr1 commented:
@reggie7 the author/publish version of peregrine is handled by a run mode - sitemap generation does not have to be concerned about versions and labels in the author/publish use case and can directly operate on the resources as before

reggie7 commented:
@reusr1 - yes, I know, but we are changing this behavior [here](https://github.com/headwirecom/peregrine-cms/pull/610/files#diff-b8050a5f9ec46eaf8bc2b91e6a646e6d1c5d4df707e8d47474af2587f3baa011R167) to explicitly use the latest version, labelled **Published**.

@GastonGonzalez we could, alternatively to what I've commented above, simply try to use the default resource for properties [here](https://github.com/headwirecom/peregrine-cms/pull/610/files#diff-b8050a5f9ec46eaf8bc2b91e6a646e6d1c5d4df707e8d47474af2587f3baa011R174) and [here](https://github.com/headwirecom/peregrine-cms/pull/610/files#diff-b8050a5f9ec46eaf8bc2b91e6a646e6d1c5d4df707e8d47474af2587f3baa011R181) instead.

GastonGonzalez commented:
@reggie7, @reusr1 and @cmrockwell - We had an offline conversion on Slack about this. On Monday, @reggie7 will adjust the behavior as follows, so that the sitemap logic reflects the correct published state for both author and publish runmodes.

1. If runmode=publish (i.e. "remote" replication service), version labels will be disabled for sitemap generation. The publish instance already contains the correct state.
2. if runmode=author (i.e. "localFS" replication service), version labels are used to generate the sitemap.

GastonGonzalez commented:
@reggie7 - The current version broke the following regression.

Pages with the page property `excludeFromSitemap=true` are being added to the sitemap on initial site replication. For example:

```
/errors/404.html
/skeleton-pages/*
```

reggie7 committed:
Fix excludeFromSitemap issue:

Pages with the page property excludeFromSitemap=true are being added to the sitemap on initial site replication. For example:
/errors/404.html
/skeleton-pages/*

reggie7 commented:
> Pages with the page property `excludeFromSitemap=true` are being added to the sitemap on initial site replication. For example:
> 
> ```
> /errors/404.html
> /skeleton-pages/*
> ```

Fixed.

reggie7 committed:
Introduce sitemap props source config

reggie7 committed:
Don't use versions in Publish

reggie7 commented:
[publish/author](https://github.com/headwirecom/peregrine-cms/pull/610#issuecomment-710127700) distinction is implemented now.

GastonGonzalez commented:
@reggie7 - This version has a critical issue which at the very least prevents new tenants from being created.

**Steps to reproduce**
1. Log in as admin
2. Create a tenant

*Observed Behavior*
1. Site is not created. Exception in the server-side logs and client-side console.

*Expected Behavior*
1. Site is created.
2. All regression below pass.

**Testing & Regression**
Please ensure that your testing involves the following:
1. New site creation and initial replication
2. Page add/modify/delete is reflected in sitemap
3. Pages marked as excluded are not in sitemap
4. Versioning

reggie7 commented:
@GastonGonzalez we'll need to discuss it in details - I can't reproduce the issue in my environment unfortunately.

GastonGonzalez commented:
@reggie7 - Please fix this issue first. Create a new branch off `develop-sling12`. Once the core copy issue is resolved we can revisit [PR 610](https://github.com/headwirecom/peregrine-cms/pull/610).

GastonGonzalez commented:
> @reggie7 - This version has a critical issue which at the very least prevents new tenants from being created.
> 
> **Steps to reproduce**
> 
> 1. Log in as admin
> 2. Create a tenant
> 
> _Observed Behavior_
> 
> 1. Site is not created. Exception in the server-side logs and client-side console.
> 
> _Expected Behavior_
> 
> 1. Site is created.
> 2. All regression below pass.
> 
> **Testing & Regression**
> Please ensure that your testing involves the following:
> 
> 1. New site creation and initial replication
> 2. Page add/modify/delete is reflected in sitemap
> 3. Pages marked as excluded are not in sitemap
> 4. Versioning

This issue shall be addressed by [issues/615](https://github.com/headwirecom/peregrine-cms/issues/615).

reggie7 committed:
Safeguard copy operation against replication properties

reggie7 committed:
Allow same parent copy

reggie7 committed:
Fix NPE

GastonGonzalez commented:
@reggie7 - Please merge your changes from changes from PR #616 so that can perform and end-to-end test before this gets merged to develop-sling12.

reggie7 committed:
Merge branch 'issues/615-copy-operations-fail-on-versioned-resources' into issues/603-update-sitemap-to-support-replicated-version

# Conflicts:
#	admin-base/core/src/main/java/com/peregrine/admin/resource/AdminResourceHandlerService.java

reggie7 committed:
Revert local-only change

reggie7 committed:
Simplify code

reggie7 committed:
Merge branch 'develop-sling12' into issues/603-update-sitemap-to-support-replicated-version

GastonGonzalez commented:
Amazing work here, @reggie7.

I went through an exhaustive test in a clean Docker instance and this feature looks great.

1. Start clean Docker instance; deploy branch
2. Log in as admin; create tenant
3. Change Site Map Structure Cache “Debounce Interval” to 30000 (30 seconds) to speed up testing.
4. Log out; log in as tenant user 
5. State of sitemap on initial tenant creation => no sitemap
6. Set domain under root template settings
7. State of sitemap after initial site setup replication
8. Pages marked for exclusion are not included from sitemap
9. Create page (p1); do not publish => not in sitemap
10. Publish page (p1) => in sitemap
11. Unpublish page (p1) => not in sitemap
12. Regression - copy p1
13. Re-publish page (p1) => in sitemap
14. Create page version (p1v2); change frequency => no change to sitemap
15. Publish page (p1v2); => frequency changed in sitemap
